### PR TITLE
IGNITE-19085 Fix waiting for indexes on recovery when building indexes

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-19085

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 1]
```diff
@@ -413,4 +392,24 @@ private interface IndexStorageAdapterFactory {
...
+     */
+    // TODO: IGNITE-19082 Needs to be redone/improved
+    public void addIndexesToWait(Collection<UUID> indexIds) {
+        for (UUID indexId : indexIds) {
+            indexesToWait.computeIfAbsent(indexId, uuid -> new CompletableFuture<>());
+        }
+    }
+
+    private void completeRegisterIndex(UUID indexId) {
+        CompletableFuture<?> indexToWaitFuture = indexesToWait.computeIfAbsent(indexId, uuid -> new CompletableFuture<>());
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:409]: It used to be `remove` but now its `computeIfAbsent`, why?

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 2]
```diff
@@ -413,4 +392,24 @@ private interface IndexStorageAdapterFactory {
         /** Creates the index decorator for given partition. */
         TableSchemaAwareIndexStorage create(int partitionId);
     }
+
+    /**
+     * Adds indexes to wait before inserting data into the table.
+     *
+     * @param indexIds Indexes Index IDs.
+     */
+    // TODO: IGNITE-19082 Needs to be redone/improved
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:401]: Did you discuss this with @AMashenkov (the author of this ticket) and that it will indeed be done under this ticket?

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 3]
```diff
@@ -361,33 +348,21 @@ public void registerSortedIndex(
     public void unregisterIndex(UUID indexId) {
         indexLockerFactories.remove(indexId);
         indexStorageAdapterFactories.remove(indexId);
+
+        CompletableFuture<?> indexToWaitFuture = indexesToWait.remove(indexId);
+
+        if (indexToWaitFuture != null) {
+            indexToWaitFuture.complete(null);
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:355]: Do we test this behavior anywhere? Shouldn't this future be completed with an exception? How does this scenario work?

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 4]
```diff
@@ -361,33 +348,21 @@ public void registerSortedIndex(
...
     }
 
     private void awaitIndexes() {
         // TODO: replace with actual call to ids supplier
-        List<UUID> indexIds = List.of(pkId()); // activeIndexIds.get();
-
         List<CompletableFuture<?>> toWait = new ArrayList<>();
 
-        for (UUID indexId : indexIds) {
-            if (indexLockerFactories.containsKey(indexId) && indexStorageAdapterFactories.containsKey(indexId)) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:373]: Why did you remove this check?

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java [hunk 5]
```diff
@@ -361,33 +348,21 @@ public void registerSortedIndex(
...
+        if (indexToWaitFuture != null) {
+            indexToWaitFuture.complete(null);
+        }
     }
 
     private void awaitIndexes() {
         // TODO: replace with actual call to ids supplier
-        List<UUID> indexIds = List.of(pkId()); // activeIndexIds.get();
-
         List<CompletableFuture<?>> toWait = new ArrayList<>();
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:361]: Just a nitpick, but you can use an array directly, without an intermediate `ArrayList`

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 6]
```diff
@@ -673,15 +675,18 @@ private CompletableFuture<?> updateAssignmentInternal(ConfigurationNotificationE
                 return failedFuture(e);
             }
 
+            TableImpl table = tablesById.get(tblId);
+
+            // TODO: IGNITE-19082 Need another way to wait for indexes
+            table.addIndexesToWait(collectTableIndexes(tablesCfg.value(), tblId));
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:681]: Why is this code located in `onUpdateAssignments` and not in table or index creation listeners?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 7]
```diff
@@ -2345,4 +2350,20 @@ private static void handleExceptionOnCleanUpTablesResources(
             nodeStoppingEx.set(true);
         }
     }
+
+    private static Collection<UUID> collectTableIndexes(TablesView tablesConfig, UUID tableId) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2354]: Another little nitpick: this method does not have to be `static`, since `tablesConfig` is a field of this class and can be removed from parameters

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 6 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:681]: Why is this code even needed? I can't understand this from the ticket description and we already populate in index-to-wait map when registering indices

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 7 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2354]: Fix it.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 5 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:361]: Just an array will not work for me, since the size of a `ConcurrentHashMap` can change at any time.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 1 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:409]: In general, it can be redone for removal.
I will do so.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 2 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:401]: Right now I'm close with the `IndexManager`, it should finish waiting for the indexes, since it will be removed in this ticket, then its author will need to take care of this problem.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 3 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:355]: No, right now the deletion of the index is not implemented correctly in general and not tested as it should, I will create a ticket for this.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 4 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:373]: For me, this check does not make sense at the moment.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 6 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:681]: Before recovery raft groups, it is important for us to know the list of indices that will be updated on recovery (in `PartitionListener`), now we are waiting only for the PK index, and we forget about the rest, which is not correct.

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 5 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:361]: ok, makes sense

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 3 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:355]: Please add a TODO here

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 4 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:373]: Can you clarify please?

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 2 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:401]: I'm just checking that the author is aware of the problem you are trying to fix and knows about this TODO

File: modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java (hunk 4 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:373]: If the index is registered, then the future of waiting for the indexes will not be or  completed, so there is no need to check the maps.

Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/TableImpl.java:373]: ok, I've read the code more carefully and looks like this check doesn't make much sense indeed

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 6 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:681]: After an in-person discussion we decided that this code should actually be moved to the table creation listeners

Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java]: Discussed personally, I think that the problem has become clearer.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 8]
```diff
@@ -1106,7 +1107,10 @@ private CompletableFuture<?> createTableLocally(long causalityToken, String name
                 partitions, clusterNodeResolver, txManager, tableStorage, txStateStorage, replicaSvc, clock);
 
         // TODO: IGNITE-16288 directIndexIds should use async configuration API
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1109]: Looks like this TODO is not relevant anymore

**Review by sashapolo — APPROVED** [review]

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 8 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1109]: Fix it

**Review by ibessonov — APPROVED** [review]
